File: js/src/function_call.ts
// Manage existing Function Calls (look-ups, polling for output, cancellation).

import { FunctionCallFromIdRequest } from "../proto/modal_proto/api";
import { type ModalClient } from "./client";
import { ControlPlaneInvocation } from "./invocation";
import { checkForRenamedParams } from "./validation";

/**
 * Service for managing {@link FunctionCall}s.
 *
 * Normally only ever accessed via the client as:
 *
 * ```typescript
 * const modal = new ModalClient();
 * const functionCall = await modal.functionCalls.fromId("123");
 * ```
 */
export class FunctionCallService {
  readonly #client: ModalClient;
  constructor(client: ModalClient) {
    this.#client = client;
  }

  /**
   * Create a new {@link FunctionCall} from ID.
   */
  async fromId(functionCallId: string): Promise<FunctionCall> {
    const resp = await this.#client.cpClient.functionCallFromId(
      FunctionCallFromIdRequest.create({ functionCallId }),
    );
    return new FunctionCall(
      this.#client,
      functionCallId,
      resp.metadata?.appId ?? "",
      resp.metadata?.functionId ?? "",
    );
  }
}

/** Optional parameters for {@link FunctionCall#get FunctionCall.get()}. */
export type FunctionCallGetParams = {
  timeoutMs?: number;
};

/** Optional parameters for {@link FunctionCall#cancel FunctionCall.cancel()}. */
export type FunctionCallCancelParams = {
  terminateContainers?: boolean;
};

/**
 * Represents a Modal FunctionCall. FunctionCalls are {@link Function_ Function} invocations with
 * a given input. They can be consumed asynchronously (see {@link FunctionCall#get FunctionCall.get()}) or cancelled
 * (see {@link FunctionCall#cancel FunctionCall.cancel()}).
 */
export class FunctionCall {
  readonly functionCallId: string;
  #client: ModalClient;
  // eslint-disable-next-line
  #appId: string;
  // eslint-disable-next-line
  #functionId: string;

  /** @ignore */
  constructor(
    client: ModalClient,
    functionCallId: string,
    appId: string,
    functionId: string,
  ) {
    this.#appId = appId;
    this.#functionId = functionId;
    this.#client = client;
    this.functionCallId = functionCallId;
  }

  /** Get the result of a FunctionCall, optionally waiting with a timeout. */
  async get(params: FunctionCallGetParams = {}): Promise<any> {
    checkForRenamedParams(params, { timeout: "timeoutMs" });

    const invocation = ControlPlaneInvocation.fromFunctionCallId(
      this.#client,
      this.functionCallId,
    );
    return invocation.awaitOutput(params.timeoutMs);
  }

  /** Cancel a running FunctionCall. */
  async cancel(params: FunctionCallCancelParams = {}) {
    const cpClient = this.#client.cpClient;

    await cpClient.functionCallCancel({
      functionCallId: this.functionCallId,
      terminateContainers: params.terminateContainers,
    });
  }
}

File: go/function_call.go
package modal

import (
	"context"
	"fmt"
	"time"

	pb "github.com/modal-labs/modal-client/go/proto/modal_proto"
)

// FunctionCallService provides FunctionCall related operations.
type FunctionCallService interface {
	FromID(ctx context.Context, functionCallID string, params *FunctionCallFromIDParams) (*FunctionCall, error)
}

// FunctionCallFromIDParams are options for FunctionCallService.FromID.
type FunctionCallFromIDParams struct{}

type functionCallServiceImpl struct{ client *Client }

// FunctionCall references a Modal Function Call. Function Calls are
// Function invocations with a given input. They can be consumed
// asynchronously (see Get()) or cancelled (see Cancel()).
type FunctionCall struct {
	FunctionCallID string
	functionID     string
	appID          string
	client         *Client
}

// FromID looks up a FunctionCall by ID.
func (s *functionCallServiceImpl) FromID(ctx context.Context, functionCallID string, params *FunctionCallFromIDParams) (*FunctionCall, error) {

	response, err := s.client.cpClient.FunctionCallFromId(ctx, pb.FunctionCallFromIdRequest_builder{FunctionCallId: functionCallID}.Build())

	if err != nil {
		return nil, fmt.Errorf("FunctionCallFromId failed: %w", err)
	}

	meta := response.GetMetadata()

	functionCall := FunctionCall{
		FunctionCallID: functionCallID,
		functionID:     meta.GetFunctionId(),
		appID:          meta.GetAppId(),
		client:         s.client,
	}
	return &functionCall, nil
}

// FunctionCallGetParams are options for getting outputs from Function Calls.
type FunctionCallGetParams struct {
	// Timeout specifies the maximum duration to wait for the output.
	// If nil, no timeout is applied. If set to 0, it will check if the function
	// call is already completed.
	Timeout *time.Duration
}

// Get waits for the output of a FunctionCall.
// If timeout > 0, the operation will be cancelled after the specified duration.
func (fc *FunctionCall) Get(ctx context.Context, params *FunctionCallGetParams) (any, error) {
	if params == nil {
		params = &FunctionCallGetParams{}
	}
	invocation := controlPlaneInvocationFromFunctionCallID(fc.client.cpClient, fc.client.logger, fc.FunctionCallID)
	return invocation.awaitOutput(ctx, params.Timeout)
}

// FunctionCallCancelParams are options for cancelling Function Calls.
type FunctionCallCancelParams struct {
	TerminateContainers bool
}

// Cancel cancels a FunctionCall.
func (fc *FunctionCall) Cancel(ctx context.Context, params *FunctionCallCancelParams) error {
	if params == nil {
		params = &FunctionCallCancelParams{}
	}
	_, err := fc.client.cpClient.FunctionCallCancel(ctx, pb.FunctionCallCancelRequest_builder{
		FunctionCallId:      fc.FunctionCallID,
		TerminateContainers: params.TerminateContainers,
	}.Build())
	if err != nil {
		return fmt.Errorf("FunctionCallCancel failed: %w", err)
	}

	return nil
}
